airbyte-integrations/connectors/destination-s3/src/test/java/io/airbyte/integrations/destination/s3/S3DestinationTest.java

```diff
@@ -20,19 +34,66 @@
 public class S3DestinationTest {
 
   private AmazonS3 s3;
+  private AmazonS3 s3NoAccess;
   private S3DestinationConfig config;
 
   @BeforeEach
   public void setup() {
+    // S3 client mock for successful operations
     s3 = mock(AmazonS3.class);
+    final InitiateMultipartUploadResult uploadResult = mock(InitiateMultipartUploadResult.class);
+    final UploadPartResult uploadPartResult = mock(UploadPartResult.class);
+    when(s3.uploadPart(any(UploadPartRequest.class))).thenReturn(uploadPartResult);
+    when(s3.initiateMultipartUpload(any(InitiateMultipartUploadRequest.class))).thenReturn(uploadResult);
+
+    // S3 client mock that throws Access Denied exception for listObjects operation
+    s3NoAccess = mock(AmazonS3.class);
+    doThrow(new AmazonS3Exception("Access Denied")).when(s3NoAccess).listObjects(any(ListObjectsRequest.class));
     config = new S3DestinationConfig(
         "fake-endpoint",
         "fake-bucket",
         "fake-bucketPath",
         "fake-region",
         "fake-accessKeyId",
         "fake-secretAccessKey",
-        null);
+        S3DestinationConfig.DEFAULT_PART_SIZE_MB, null, s3);
+  }
+
+  @Test
+  public void checksS3NoListObjectPermission() {
+    final S3Destination destinationFail = new S3Destination(new S3DestinationConfigFactory () {
+      public S3DestinationConfig getS3DestinationConfig(final JsonNode config) {
+        return new S3DestinationConfig(
+            "fake-endpoint",
+            "fake-bucket",
+            "fake-bucketPath",
+            "fake-region",
+            "fake-accessKeyId",
+            "fake-secretAccessKey",
+            S3DestinationConfig.DEFAULT_PART_SIZE_MB,
+            null, s3NoAccess);
+      }
+    });
+    AirbyteConnectionStatus status = destinationFail.check(null);
+    assertEquals(Status.FAILED, status.getStatus(), "Connection check should have failed");
+    assertTrue(status.getMessage().indexOf("Access Denied") > 0, "Connection check returned wrong failure message");
+
+    // Test that check succeeds when IAM user has listObjects permission
```

Comment:
ideally each unit test verifies a single well defined scenario, so we should split these into two unit tests

Reply:
done